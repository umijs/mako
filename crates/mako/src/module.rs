use std::collections::HashSet;
use std::fmt::{Debug, Formatter};
use std::path::PathBuf;
use std::sync::Arc;

use anyhow::{anyhow, Result};
use bitflags::bitflags;
use pathdiff::diff_paths;
use serde::Serialize;
use swc_core::common::{Span, DUMMY_SP};
use swc_core::ecma::ast::{
    BlockStmt, ExportSpecifier, FnExpr, Function, ImportDecl, ImportSpecifier, Module as SwcModule,
    ModuleExportName, NamedExport,
};
use swc_core::ecma::utils::quote_ident;

use crate::ast::css_ast::CssAst;
use crate::ast::file::{win_path, File};
use crate::ast::js_ast::JsAst;
use crate::build::analyze_deps::AnalyzeDepsResult;
use crate::compiler::Context;
use crate::config::ModuleIdStrategy;
use crate::resolve::ResolverResource;

pub type Dependencies = HashSet<Dependency>;

#[derive(Debug, Clone, Hash, Eq, PartialEq)]
pub struct Dependency {
    pub source: String,
    pub resolve_as: Option<String>,
    pub resolve_type: ResolveType,
    pub order: usize,
    pub span: Option<Span>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum ModuleSystem {
    CommonJS,
    ESModule,
    Custom,
}

bitflags! {
    #[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Default)]
    pub struct ResolveTypeFlags: u16 {
        const Sync  = 1;
        const Async = 1<<2;
    }

    #[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Default)]
    pub struct ImportType: u16 {
        const Default = 1;
        const Named = 1<<2;
        const Namespace = 1<<3;
        const SideEffect = 1<<4 ;
    }

    #[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Default)]
    pub struct NamedExportType: u16 {
        const Named = 1;
        const Default = 1<<2;
        const Namespace = 1<<3;
    }
}

impl From<&ResolveType> for ResolveTypeFlags {
    fn from(value: &ResolveType) -> Self {
        match value {
            ResolveType::DynamicImport(_) | ResolveType::Worker(_) => Self::Async,
            _ => Self::Sync,
        }
    }
}

impl From<&ImportDecl> for ImportType {
    fn from(decl: &ImportDecl) -> Self {
        if decl.specifiers.is_empty() {
            ImportType::SideEffect
        } else {
            let mut import_type = ImportType::empty();
            for specifier in &decl.specifiers {
                match specifier {
                    ImportSpecifier::Named(_) => {
                        import_type |= ImportType::Named;
                    }
                    ImportSpecifier::Default(_) => {
                        import_type |= ImportType::Default;
                    }
                    ImportSpecifier::Namespace(_) => {
                        import_type |= ImportType::Namespace;
                    }
                }
            }
            import_type
        }
    }
}

impl From<&NamedExportType> for ImportType {
    fn from(value: &NamedExportType) -> Self {
        let mut res = Self::empty();
        value.iter().for_each(|b| match b {
            NamedExportType::Default => res.insert(Self::Default),
            NamedExportType::Named => res.insert(Self::Named),
            NamedExportType::Namespace => res.insert(Self::Namespace),
            _ => {}
        });
        res
    }
}

impl From<&NamedExport> for NamedExportType {
    fn from(decl: &NamedExport) -> Self {
        let mut res = Self::empty();

        decl.specifiers
            .iter()
            .for_each(|specifier| match specifier {
                ExportSpecifier::Namespace(_) => res.insert(Self::Namespace),
                ExportSpecifier::Default(_) => res.insert(Self::Default),
                ExportSpecifier::Named(named) => {
                    if let ModuleExportName::Ident(orig) = &named.orig
                        && orig.sym.eq("default")
                    {
                        res.insert(Self::Default);
                    } else {
                        res.insert(Self::Named);
                    }
                }
            });

        res
    }
}

#[derive(Eq, Hash, PartialEq, Serialize, Debug, Clone, Default)]
pub struct ImportOptions {
    pub chunk_name: Option<String>,
    pub ignore: bool,
}

impl ImportOptions {
    pub fn get_chunk_name(&self) -> &Option<String> {
        &self.chunk_name
    }
}

#[derive(Eq, Hash, PartialEq, Serialize, Debug, Clone)]
pub enum ResolveType {
    Import(ImportType),
    ExportNamed(NamedExportType),
    ExportAll,
    Require,
    DynamicImport(ImportOptions),
    Css,
    Worker(ImportOptions),
}

impl ResolveType {
    pub fn same_enum(&self, other: &Self) -> bool {
        match (self, other) {
            (Self::Import(_), Self::Import(_)) => true,
            (_, _) => self == other,
        }
    }
}

impl ResolveType {
    pub fn is_esm(&self) -> bool {
        self.is_sync_esm() || self.is_dynamic_esm()
    }

    pub fn is_sync_esm(&self) -> bool {
        matches!(
            self,
            ResolveType::Import(_) | ResolveType::ExportNamed(_) | ResolveType::ExportAll
        )
    }

    pub fn is_dynamic_esm(&self) -> bool {
        matches!(self, ResolveType::DynamicImport(_))
    }
}

#[derive(Debug, Clone)]
pub struct ModuleInfo {
    pub ast: ModuleAst,
    pub file: File,
    pub deps: AnalyzeDepsResult,
    pub external: Option<String>,
    pub raw: String,
    pub raw_hash: u64,
    /// Modules with top-level-await
    pub top_level_await: bool,
    /// The top-level-await module must be an async module, in addition, for example, wasm is also an async module
    /// The purpose of distinguishing top_level_await and is_async is to adapt to runtime_async
    pub is_async: bool,
    pub is_ignored: bool,
    pub resolved_resource: Option<ResolverResource>,
    /// The transformed source map chain of this module
    pub source_map_chain: Vec<Vec<u8>>,
    pub module_system: ModuleSystem,
}

impl Default for ModuleInfo {
    fn default() -> Self {
        Self {
            module_system: ModuleSystem::CommonJS,
            ast: ModuleAst::None,
            file: Default::default(),
            deps: Default::default(),
            external: None,
            raw: "".to_string(),
            raw_hash: 0,
            top_level_await: false,
            is_async: false,
            resolved_resource: None,
            source_map_chain: vec![],
            is_ignored: false,
        }
    }
}

fn md5_hash(source_str: &str, lens: usize) -> String {
    format!("{:x}", md5::compute(source_str))
        .chars()
        .take(lens)
        .collect::<String>()
}

pub fn generate_module_id(origin_module_id: &str, context: &Arc<Context>) -> String {
    match context.config.module_id_strategy {
        ModuleIdStrategy::Hashed => md5_hash(origin_module_id, 8),
        ModuleIdStrategy::Named => {
            // readable ids for debugging usage
            let absolute_path = PathBuf::from(origin_module_id);
            let relative_path = diff_paths(&absolute_path, &context.root).unwrap_or(absolute_path);
            win_path(relative_path.to_str().unwrap())
        }
        ModuleIdStrategy::Numeric => {
            let numeric_ids_map = context.numeric_ids_map.read().unwrap();
            if let Some(numeric_id) = numeric_ids_map.get(origin_module_id) {
                numeric_id.to_string()
            } else {
                md5_hash(origin_module_id, 8)
            }
        }
    }
}

pub fn relative_to_root(module_path: &String, root: &PathBuf) -> String {
    let absolute_path = PathBuf::from(module_path);
    let relative_path = diff_paths(&absolute_path, root).unwrap_or(absolute_path);
    // diff_paths result always starts with ".."/"." or not
    if relative_path.starts_with("..") || relative_path.starts_with(".") {
        relative_path.to_string_lossy().to_string()
    } else {
        PathBuf::from(".")
            .join(relative_path)
            .to_string_lossy()
            .to_string()
    }
}

#[derive(Clone, Eq, PartialEq, Hash, Debug)]
pub struct ModuleId {
    pub id: String,
}

impl Ord for ModuleId {
    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
        self.id.cmp(&other.id)
    }
}

impl PartialOrd for ModuleId {
    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
        Some(self.cmp(other))
    }
}

impl ModuleId {
    // we use absolute path as module id now
    pub fn new(id: String) -> Self {
        Self { id }
    }

    pub fn generate(&self, context: &Arc<Context>) -> String {
        // TODO: 如果是 Hashed 的话，stats 拿不到原始的 chunk_id
        generate_module_id(&self.id, context)
    }

    pub fn from_path(path_buf: PathBuf) -> Self {
        Self {
            id: path_buf.to_string_lossy().to_string(),
        }
    }

    // FIXME: 这里暂时直接通过 module_id 转换为 path，后续如果改了逻辑要记得改
    pub fn to_path(&self) -> PathBuf {
        PathBuf::from(self.id.clone())
    }
}

impl From<String> for ModuleId {
    fn from(id: String) -> Self {
        Self { id }
    }
}

impl From<&str> for ModuleId {
    fn from(id: &str) -> Self {
        Self { id: id.to_string() }
    }
}

impl From<PathBuf> for ModuleId {
    fn from(path: PathBuf) -> Self {
        Self {
            id: path.to_string_lossy().to_string(),
        }
    }
}

#[derive(Debug, Clone)]
pub enum ModuleAst {
    Script(JsAst),
    Css(CssAst),
    None,
}

impl ModuleAst {
    #[allow(dead_code)]
    pub fn as_script(&self) -> Option<&JsAst> {
        match self {
            ModuleAst::Script(ast) => Some(ast),
            _ => None,
        }
    }

    pub fn script_mut(&mut self) -> Option<&mut JsAst> {
        match self {
            ModuleAst::Script(ast) => Some(ast),
            _ => None,
        }
    }

    pub fn as_script_ast(&self) -> &SwcModule {
        if let Self::Script(script) = self {
            &script.ast
        } else {
            panic!("ModuleAst is not Script")
        }
    }

    pub fn as_script_ast_mut(&mut self) -> &mut SwcModule {
        if let Self::Script(script) = self {
            &mut script.ast
        } else {
            panic!("ModuleAst is not Script")
        }
    }

    pub fn as_css_mut(&mut self) -> &mut CssAst {
        if let Self::Css(css) = self {
            css
        } else {
            panic!("ModuleAst is not Css")
        }
    }

    pub fn as_script_mut(&mut self) -> &mut JsAst {
        if let Self::Script(script) = self {
            script
        } else {
            panic!("ModuleAst is not Css")
        }
    }
}

#[derive(PartialEq, Eq, Debug)]
pub enum ModuleType {
    Script,
    Css,
    Raw,
    PlaceHolder,
}

#[derive(Clone)]
pub struct Module {
    pub id: ModuleId,
    pub is_entry: bool,
    pub info: Option<ModuleInfo>,
    pub side_effects: bool,
}

impl Module {
    pub fn new(id: ModuleId, is_entry: bool, info: Option<ModuleInfo>) -> Self {
        Self {
            id,
            is_entry,
            info,
            side_effects: is_entry,
        }
    }

    pub fn as_script(&self) -> Option<&JsAst> {
        self.info.as_ref().and_then(|i| i.ast.as_script())
    }

    pub fn as_mut_script(&mut self) -> Option<&mut JsAst> {
        self.info.as_mut().and_then(|i| i.ast.script_mut())
    }

    pub fn set_info(&mut self, info: Option<ModuleInfo>) {
        self.info = info;
    }

    pub fn is_external(&self) -> bool {
        self.info
            .as_ref()
            .map_or(false, |info| info.external.is_some())
    }

    pub fn is_placeholder(&self) -> bool {
        self.get_module_type() == ModuleType::PlaceHolder
    }

    pub fn get_module_type(&self) -> ModuleType {
        self.info
            .as_ref()
            .map_or(ModuleType::PlaceHolder, |info| match info.ast {
                ModuleAst::Script(_) => ModuleType::Script,
                ModuleAst::Css(_) => ModuleType::Css,
                ModuleAst::None => ModuleType::Raw,
            })
    }

    pub fn get_module_size(&self) -> usize {
        self.info
            .as_ref()
            .map_or(0, |info| info.raw.as_bytes().len())
    }

    // wrap module stmt into a function
    // eg:
    // function(module, exports, require) {
    //   module stmt..
    // }
    pub fn to_module_fn_expr(&self) -> Result<FnExpr> {
        match &self.info.as_ref().unwrap().ast {
            ModuleAst::Script(script) => {
                let mut stmts = Vec::new();

                for n in script.ast.body.iter() {
                    match n.as_stmt() {
                        None => {
                            return Err(anyhow!(
                                "Error: not a stmt found in {:?}, ast: {:?}",
                                self.id.id,
                                n,
                            ));
                        }
                        Some(stmt) => {
                            stmts.push(stmt.clone());
                        }
                    }
                }

                let func = Function {
                    span: DUMMY_SP,
                    ctxt: Default::default(),
                    params: vec![
                        quote_ident!("module").into(),
                        quote_ident!("exports").into(),
                        quote_ident!("__mako_require__").into(),
                    ],
                    decorators: vec![],
                    body: Some(BlockStmt {
                        span: DUMMY_SP,
                        ctxt: Default::default(),
                        stmts,
                    }),
                    is_generator: false,
                    is_async: false,
                    type_params: None,
                    return_type: None,
                };
                Ok(FnExpr {
                    ident: None,
                    function: func.into(),
                })
            }
            // TODO: css modules will be removed in the future
            ModuleAst::Css(_) => Ok(empty_module_fn_expr()),
            ModuleAst::None => Err(anyhow!("ModuleAst::None({}) cannot concert", self.id.id)),
        }
    }
}

impl Debug for Module {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(f, "id={}({:?})", self.id.id, self.get_module_type())
    }
}

fn empty_module_fn_expr() -> FnExpr {
    let func = Function {
        ctxt: Default::default(),
        span: DUMMY_SP,
        params: vec![
            quote_ident!("module").into(),
            quote_ident!("exports").into(),
            quote_ident!("__mako_require__").into(),
        ],
        decorators: vec![],
        body: Some(BlockStmt {
            ctxt: Default::default(),
            span: DUMMY_SP,
            stmts: vec![],
        }),
        is_generator: false,
        is_async: false,
        type_params: None,
        return_type: None,
    };
    FnExpr {
        ident: None,
        function: func.into(),
    }
}
