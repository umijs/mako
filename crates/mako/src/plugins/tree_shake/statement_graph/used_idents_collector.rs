use std::collections::HashSet;

use swc_core::ecma::ast::Ident;
use swc_core::ecma::visit::Visit;

pub struct UsedIdentsCollector {
    pub used_idents: HashSet<String>,
}

impl UsedIdentsCollector {
    pub fn new() -> Self {
        Self {
            used_idents: HashSet::new(),
        }
    }
}

impl Visit for UsedIdentsCollector {
    fn visit_ident(&mut self, ident: &Ident) {
        self.used_idents.insert(ident.to_string());
    }
}
