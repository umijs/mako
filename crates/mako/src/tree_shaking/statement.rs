use std::collections::HashSet;

#[derive(Debug, Clone)]
pub enum ImportSpecifier {
    // import * as foo from 'foo';
    Namespace(String),
    // import { foo, bar, default as zoo } from 'foo';
    Named {
        local: String,
        imported: Option<String>,
    },
    // import foo from 'foo';
    Default(String),
}

#[derive(Debug, Clone)]
pub enum ExportSpecifier {
    // export * from 'foo';
    All,
    // export { foo, bar, default as zoo } from 'foo';
    Named {
        local: String,
        // "as zoo" is exported
        exported: Option<String>,
    },
    // export default xxx;
    Default,
    // export * as foo from 'foo';
    Namespace(String),
}

#[derive(Debug, Clone)]
pub struct ImportInfo {
    pub source: String,
    pub specifiers: Vec<ImportSpecifier>,
    pub stmt_id: StatementId,
}

#[derive(Debug, Clone)]
pub struct ExportInfo {
    pub source: Option<String>,
    pub specifiers: Vec<ExportSpecifier>,
    pub stmt_id: StatementId,
}

pub type StatementId = usize;

#[derive(Debug, Clone)]
pub struct ImportStatement {
    pub id: StatementId,
    pub info: ImportInfo,
    pub is_self_executed: bool,
    pub defined_ident: HashSet<String>,
}

#[derive(Debug, Clone)]
pub struct ExportStatement {
    pub id: StatementId,
    pub info: ExportInfo,
    pub defined_ident: HashSet<String>,
    pub used_ident: HashSet<String>,
}

#[derive(Debug, Clone)]
pub enum StatementType {
    Import(ImportStatement),
    Export(ExportStatement),
    Stmt {
        id: StatementId,
        defined_ident: HashSet<String>,
        used_ident: HashSet<String>,
        is_self_executed: bool,
    },
}

impl StatementType {
    pub fn get_id(&self) -> StatementId {
        match self {
            StatementType::Import(ImportStatement { id, .. }) => *id,
            StatementType::Export(ExportStatement { id, .. }) => *id,
            StatementType::Stmt { id, .. } => *id,
        }
    }

    pub fn get_is_self_executed(&self) -> bool {
        match self {
            StatementType::Import(ImportStatement {
                is_self_executed, ..
            }) => *is_self_executed,
            StatementType::Export(ExportStatement {
                defined_ident: _, ..
            }) => false,
            StatementType::Stmt {
                is_self_executed, ..
            } => *is_self_executed,
        }
    }

    pub fn get_defined_ident(&self) -> &HashSet<String> {
        match self {
            StatementType::Import(ImportStatement { defined_ident, .. }) => defined_ident,
            StatementType::Export(ExportStatement { defined_ident, .. }) => defined_ident,
            StatementType::Stmt { defined_ident, .. } => defined_ident,
        }
    }
    pub fn get_used_ident(&self) -> Option<&HashSet<String>> {
        match self {
            StatementType::Import(ImportStatement { .. }) => None,
            StatementType::Export(ExportStatement { used_ident, .. }) => Option::Some(used_ident),
            StatementType::Stmt { used_ident, .. } => Option::Some(used_ident),
        }
    }
}
