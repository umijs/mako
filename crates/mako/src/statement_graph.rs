use std::collections::{HashMap, HashSet};
use std::fmt;

use petgraph::stable_graph::{NodeIndex, StableDiGraph};
use petgraph::visit::{EdgeRef, IntoEdgeReferences};
use swc_ecma_ast::Module;

use crate::analyze_statement::analyze_statement;
use crate::statement::{StatementId, StatementType};

#[derive(Debug, Clone, Eq, PartialEq)]
pub struct StatementGraphEdge {
    pub ident: HashSet<String>,
}

#[derive(Debug)]
pub struct StatementGraph {
    graph: StableDiGraph<StatementType, StatementGraphEdge>,
    id_index_map: HashMap<StatementId, NodeIndex>,
}

/**
 * 声明语句的依赖关系图
 */
impl StatementGraph {
    pub fn new(module: &Module) -> Self {
        let mut graph = StableDiGraph::new();
        let mut id_index_map = HashMap::new();

        // 只分析 body 顶层的声明语句
        for (index, statement) in module.body.iter().enumerate() {
            let node = graph.add_node(analyze_statement(index, statement));
            id_index_map.insert(index, node);
        }

        let mut graph = Self {
            graph,
            id_index_map,
        };

        graph.init_graph_edge();

        graph
    }

    pub fn empty() -> Self {
        Self {
            graph: StableDiGraph::new(),
            id_index_map: HashMap::new(),
        }
    }

    pub fn get_statements(&self) -> Vec<&StatementType> {
        self.graph.node_indices().map(|i| &self.graph[i]).collect()
    }

    pub fn get_statement(&self, id: &StatementId) -> &StatementType {
        let node = self.id_index_map.get(id).unwrap();
        &self.graph[*node]
    }

    pub fn get_dependencies(
        &self,
        id: &StatementId,
    ) -> Vec<(&StatementType, HashSet<String>, StatementId)> {
        let node = self.id_index_map.get(id).unwrap();
        self.graph
            .neighbors(*node)
            .map(|i| {
                let edge = self.graph.find_edge(*node, i).unwrap();
                let edge = self.graph.edge_weight(edge).unwrap();
                let statement = self.graph.node_weight(i).unwrap();
                let id = statement.get_id();
                (&self.graph[i], edge.ident.clone(), id)
            })
            .collect()
    }

    pub fn add_edge(&mut self, from: StatementId, to: StatementId, ident: HashSet<String>) {
        let from_node = self.id_index_map.get(&from).unwrap();
        let to_node = self.id_index_map.get(&to).unwrap();

        if let Some(edge) = self.graph.find_edge(*from_node, *to_node) {
            let edge = self.graph.edge_weight_mut(edge).unwrap();

            edge.ident.extend(ident);
            return;
        }

        self.graph
            .add_edge(*from_node, *to_node, StatementGraphEdge { ident });
    }

    fn init_graph_edge(&mut self) {
        let mut edges_to_add = Vec::new();
        for statement in self.get_statements() {
            for def_statement in self.get_statements() {
                let mut deps_ident = HashSet::new();
                for def_ident in def_statement.get_defined_ident() {
                    if let Some(used_ident) = statement.get_used_ident() {
                        if used_ident.contains(def_ident) {
                            deps_ident.insert(def_ident.clone());
                        }
                    }
                }
                if !deps_ident.is_empty() {
                    edges_to_add.push((statement.get_id(), def_statement.get_id(), deps_ident));
                }
            }
        }

        for (from, to, ident) in edges_to_add {
            self.add_edge(from, to, ident);
        }
    }
}

impl fmt::Display for StatementGraph {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        let mut nodes = self
            .graph
            .node_weights()
            .map(|node| {
                let id = node.get_id();
                match node {
                    StatementType::Import(import) => (
                        id,
                        format!(
                            "import {:?} - {:?}",
                            &stable_display_hashset(&import.defined_ident),
                            &import.is_self_executed
                        ),
                    ),
                    StatementType::Export(export) => (
                        id,
                        format!(
                            "export {:?} - {:?}",
                            &stable_display_hashset(&export.defined_ident),
                            &stable_display_hashset(&export.used_ident)
                        ),
                    ),
                    StatementType::Stmt {
                        id,
                        defined_ident,
                        used_ident,
                        is_self_executed,
                    } => (
                        *id,
                        format!(
                            "stmt {:?} - {:?} - {:?}",
                            &stable_display_hashset(defined_ident),
                            &stable_display_hashset(used_ident),
                            is_self_executed
                        ),
                    ),
                }
            })
            .collect::<Vec<_>>();
        let mut references = self
            .graph
            .edge_references()
            .map(|edge| {
                let source = &self.graph[edge.source()].get_id();
                let target = &self.graph[edge.target()].get_id();
                format!("{} -> {}", source, target)
            })
            .collect::<Vec<_>>();
        nodes.sort_by_key(|id| id.0);
        references.sort_by_key(|id| id.to_string());
        write!(
            f,
            "graph\n nodes:{:?}\n \n references:{:?}",
            &nodes, &references
        )
    }
}

fn stable_display_hashset<T: ToString>(a: &HashSet<T>) -> Vec<&T> {
    let mut a = a.iter().collect::<Vec<_>>();
    a.sort_by_key(|id| id.to_string());
    a
}
