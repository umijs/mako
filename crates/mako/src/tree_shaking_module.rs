use std::collections::{HashMap, HashSet};
use std::{fmt, vec};

use tracing::debug;

use crate::module::{Module, ModuleId};
use crate::statement::{
    ExportSpecifier, ExportStatement, ImportStatement, StatementId, StatementType,
};
use crate::statement_graph::StatementGraph;

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub enum UsedIdent {
    /// Local ident
    SwcIdent(String),
    /// Default ident
    Default,
    /// This ident is used and may be exported from other module
    InExportAll(String),
    /// All idents is used and may be exported from other module
    ExportAll,
}

impl ToString for UsedIdent {
    fn to_string(&self) -> String {
        match self {
            UsedIdent::SwcIdent(ident) => ident.to_string(),
            UsedIdent::Default => "default".to_string(),
            UsedIdent::InExportAll(ident) => ident.to_string(),
            UsedIdent::ExportAll => "*".to_string(),
        }
    }
}

/**
 * 当前模块被用到的exports
 */
#[derive(Debug, Clone)]
pub enum UsedExports {
    All,
    Partial(Vec<String>),
}

impl UsedExports {
    pub fn add_used_export(&mut self, used_export: &dyn ToString) {
        match self {
            UsedExports::All => {
                *self = UsedExports::All;
            }
            UsedExports::Partial(self_used_exports) => {
                self_used_exports.push(used_export.to_string())
            }
        }
    }

    pub fn is_empty(&self) -> bool {
        match self {
            UsedExports::All => false,
            UsedExports::Partial(self_used_exports) => self_used_exports.is_empty(),
        }
    }
}

pub struct UsedIdentHashMap(HashMap<StatementId, HashSet<UsedIdent>>);
impl From<UsedIdentHashMap> for Vec<(StatementId, HashSet<UsedIdent>)> {
    fn from(val: UsedIdentHashMap) -> Self {
        let mut vec = val.0.into_iter().collect::<Vec<_>>();
        vec.sort_by_key(|a| a.0);
        vec
    }
}

#[derive(Debug)]
pub struct TreeShakingModule {
    pub id: ModuleId,
    pub used_exports: UsedExports,
    pub side_effects: bool,
    statement_graph: StatementGraph,
}

impl TreeShakingModule {
    pub fn new(module: &Module) -> Self {
        let ast = &module.info.as_ref().unwrap().ast;
        let statement_graph = match ast {
            crate::module::ModuleAst::Script(module) => StatementGraph::new(&module.ast),
            crate::module::ModuleAst::Css(_) => StatementGraph::empty(),
            crate::module::ModuleAst::None => StatementGraph::empty(),
        };

        let used_exports = if module.side_effects {
            UsedExports::All
        } else {
            UsedExports::Partial(vec![])
        };

        Self {
            id: module.id.clone(),
            used_exports,
            side_effects: module.side_effects,
            statement_graph,
        }
    }

    #[allow(dead_code)]
    fn get_statements(&self) -> Vec<&StatementType> {
        self.statement_graph.get_statements()
    }

    pub fn get_statement(&self, id: &StatementId) -> &StatementType {
        self.statement_graph.get_statement(id)
    }

    pub fn imports(&self) -> Vec<ImportStatement> {
        let mut imports: Vec<ImportStatement> = vec![];
        for statement in self.statement_graph.get_statements() {
            if let StatementType::Import(statement) = &statement {
                imports.push(statement.clone());
            }
        }
        imports
    }

    pub fn exports(&self) -> Vec<ExportStatement> {
        let mut exports: Vec<ExportStatement> = vec![];
        for statement in self.statement_graph.get_statements() {
            if let StatementType::Export(statement) = &statement {
                exports.push(statement.clone());
            }
        }
        exports
    }

    /**
     * 获取使用到的所有导出的 statement
     */
    pub fn get_used_export_statement(&self) -> UsedIdentHashMap {
        let used_exports_ident = self.get_used_export_ident();
        let mut stmt_used_ident_map: HashMap<StatementId, HashSet<UsedIdent>> = HashMap::new();

        for (used_ident, stmt_id) in used_exports_ident {
            let used_idents: &mut HashSet<UsedIdent> =
                stmt_used_ident_map.entry(stmt_id).or_default();
            used_idents.insert(used_ident);
        }

        for stmt in self.statement_graph.get_statements() {
            if stmt.get_is_self_executed() {
                stmt_used_ident_map.entry(stmt.get_id()).or_default();

                for (dep_statement, referred_idents) in
                    self.statement_graph.get_dependencies(&stmt.get_id())
                {
                    let used_idents = stmt_used_ident_map
                        .entry(dep_statement.get_id())
                        .or_default();
                    used_idents.extend(referred_idents.into_iter().map(UsedIdent::SwcIdent));
                }
            }
        }

        UsedIdentHashMap(stmt_used_ident_map)
    }

    /**
     * 当前模块内到处的 identifiers
     */
    pub fn get_used_export_ident(&self) -> Vec<(UsedIdent, usize)> {
        debug!("used_exports {:?}", &self.used_exports);
        match &self.used_exports {
            UsedExports::All => {
                // all exported identifiers are used
                let mut used_ident = vec![];

                for export_statement in self.exports() {
                    for sp in export_statement.info.specifiers {
                        match sp {
                            ExportSpecifier::Default => {
                                used_ident.push((UsedIdent::Default, export_statement.id));
                            }
                            ExportSpecifier::Named { local, .. } => {
                                used_ident.push((
                                    UsedIdent::SwcIdent(local.clone()),
                                    export_statement.id,
                                ));
                            }
                            ExportSpecifier::Namespace(ns) => {
                                used_ident
                                    .push((UsedIdent::SwcIdent(ns.clone()), export_statement.id));
                            }
                            ExportSpecifier::All(_) => {
                                used_ident.push((UsedIdent::ExportAll, export_statement.id));
                            }
                        }
                    }
                }

                used_ident
            }
            UsedExports::Partial(idents) => {
                let mut used_ident = vec![];

                for ident in idents {
                    let export_statements = self.exports().into_iter().find(|export_statement| {
                        export_statement.info.specifiers.iter().any(|sp| match sp {
                            ExportSpecifier::Default => ident == "default",
                            ExportSpecifier::Named { local, exported } => {
                                let exported_ident = if let Some(exported) = exported {
                                    exported
                                } else {
                                    local
                                };

                                Self::is_same_ident(ident, exported_ident)
                            }
                            ExportSpecifier::Namespace(ns) => Self::is_same_ident(ident, ns),
                            ExportSpecifier::All(_) => false,
                        })
                    });

                    if let Some(export_statement) = export_statements {
                        for sp in export_statement.info.specifiers {
                            match sp {
                                ExportSpecifier::Default => {
                                    if ident == "default" {
                                        used_ident.push((UsedIdent::Default, export_statement.id));
                                    }
                                }
                                ExportSpecifier::Named { local, exported } => {
                                    if let Some(exported) = exported {
                                        if Self::is_same_ident(ident, &exported) {
                                            used_ident.push((
                                                UsedIdent::SwcIdent(local.clone()),
                                                export_statement.id,
                                            ));
                                        }
                                    } else if Self::is_same_ident(ident, &local) {
                                        used_ident.push((
                                            UsedIdent::SwcIdent(local.clone()),
                                            export_statement.id,
                                        ));
                                    }
                                }
                                ExportSpecifier::Namespace(ns) => {
                                    if Self::is_same_ident(ident, &ns) {
                                        used_ident.push((
                                            UsedIdent::SwcIdent(ns.clone()),
                                            export_statement.id,
                                        ));
                                    }
                                }
                                ExportSpecifier::All(_) => unreachable!(),
                            }
                        }
                    } else {
                        for export_statement in self.exports() {
                            if export_statement
                                .info
                                .specifiers
                                .iter()
                                .any(|sp| matches!(sp, ExportSpecifier::All(_)))
                            {
                                used_ident.push((
                                    UsedIdent::InExportAll(ident.to_string()),
                                    export_statement.id,
                                ));
                            }
                        }
                    }
                }

                used_ident
            }
        }
    }

    fn is_same_ident(ident1: &str, ident2: &str) -> bool {
        let split1 = ident1.split('#').collect::<Vec<_>>();
        let split2 = ident2.split('#').collect::<Vec<_>>();

        if split1.len() == 2 && split2.len() == 2 {
            split1[0] == split2[0] && split1[1] == split2[1]
        } else {
            split1[0] == split2[0]
        }
    }
}

impl fmt::Display for TreeShakingModule {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        self.statement_graph.fmt(f)
    }
}

#[cfg(test)]
mod tests {
    use std::path::PathBuf;

    use super::TreeShakingModule;
    use crate::test_helper::create_mock_module;
    use crate::{assert_debug_snapshot, assert_display_snapshot};

    #[test]
    fn test_tree_shaking_module() {
        let module = create_mock_module(
            PathBuf::from("/path/to/test"),
            r#"
import { x } from 'foo';
import 'bar';
const f0 = 1;
export const f1 = 1;
export const f2 = x;
"#,
        );
        let tree_shaking_module = TreeShakingModule::new(&module);
        assert_debug_snapshot!(&tree_shaking_module.get_statements());
        assert_eq!(tree_shaking_module.exports().len(), 2);
        assert_eq!(tree_shaking_module.imports().len(), 2);
        assert_display_snapshot!(&tree_shaking_module);
    }

    #[test]
    fn test_class_func() {
        let module = create_mock_module(
            PathBuf::from("/path/to/test"),
            r#"
import { x } from 'foo';

export const f3 = x;
export const f1 = 1;

if (true) {
    const f1 = x;

    {
        const f2 = 1;
    }

    class Foo {
        constructor() {
            x;
            f1;
            f3;
        }
    }
}

"#,
        );
        let tree_shaking_module = TreeShakingModule::new(&module);
        assert_display_snapshot!(&tree_shaking_module);
    }
}
