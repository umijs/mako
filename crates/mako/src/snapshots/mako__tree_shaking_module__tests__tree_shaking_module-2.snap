---
source: crates/mako/src/tree_shaking_module.rs
expression: "&tree_shaking_module"
---
graph
 nodes:[(0, "import [\"x#0\"] - false"), (1, "import [] - true"), (2, "stmt [\"f0#0\"] - [] - false"), (3, "export [\"f1#0\"] - []"), (4, "export [\"f2#0\"] - [\"x#0\"]")]
 
 references:["4 -> 0"]
